gateway/blocks_gateway.go

```diff
@@ -0,0 +1,484 @@
+		return ContentPathMetadata{}, nil, NewErrorResponse(fmt.Errorf("can only do range requests on files, but did not get a file"), http.StatusBadRequest)
+	}
+
+	return md, f, nil
+}
+
+func (api *BlocksGateway) GetAll(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	return md, n, nil
+}
+
+func (api *BlocksGateway) GetBlock(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.File, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, files.NewBytesFile(nd.RawData()), nil
+}
+
+func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	rootCodec := nd.Cid().Prefix().GetCodec()
+	if rootCodec != uint64(mc.DagPb) {
+		return md, files.NewBytesFile(nd.RawData()), nil
+	}
+
+	// TODO: We're not handling non-UnixFS dag-pb. There's a bit of a discrepancy between what we want from a HEAD request and a Resolve request here and we're using this for both
+	fileNode, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, fileNode, nil
+}
+
+func (api *BlocksGateway) GetCAR(ctx context.Context, path ImmutablePath) (ContentPathMetadata, io.ReadCloser, <-chan error, error) {
+	// Same go-car settings as dag.export command
+	store := dagStore{api: api, ctx: ctx}
+
+	// TODO: When switching to exposing path blocks we'll want to add these as well
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	rootCid := lastSeg.Cid()
+
+	// TODO: support selectors passed as request param: https://github.com/ipfs/kubo/issues/8769
```

Comment:
🤷 we can. Adding optional varargs isn't a breaking change so it doesn't really matter. Also, even if it is breaking that might even be a good thing since we'd want implementations to handle that case instead of oversending data back.

Note: Would we treat those kinds of requests like HTTP Range Requests where the caller isn't obligated to honor them or would they need to error with "not supported" or at least say what they support with some kind of OPTIONS request?

Reply:
### On breaking 

Ack, I'm ok with adding options later: realized we will have to add TTL to GetDNSLinkRecord at some point too (https://github.com/ipfs/kubo/issues/1818#issuecomment-1015849462), breaking is unavoidable.

### On liberal vs strict  approach to requests

Since the project's inception, Gateway did not error on unknown query parameters – unknown parameters are simply ignored. https://ipfs.io/ipfs/bafkqaaa?format=car&bytes=0-1024 does not error today – you get the full DAG.

This means we will fight a losing battle trying to enforce strict error handling behavior across all implementations and deployments. Instead, provide modern clients with the ability to probe and detect which parameters are supported via `OPTIONS /ipfs/cid?format=car`  that returns HTTP 204 with  `Ipfs-Gateway-Features: bytes, depth, filename` header. 
If there is no header, the client knows they deal with a legacy/basic gateway that has no smarts.